pkg/gitprovider/bitbucketserver.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	branchList, err := bitbucketv1.GetBranchesResponse(branches)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, branch := range branchList {
+		response = append(response, &GitBranch{
+			Name: branch.DisplayID,
+			Sha:  branch.LatestCommit,
+		})
+	}
+
+	return response, nil
+}
+
+func (g *BitbucketServerGitProvider) GetRepoPRs(repositoryId string, namespaceId string) ([]*GitPullRequest, error) {
+	client := g.getApiClient()
+	var response []*GitPullRequest
+
+	bitbucketDCClient, ok := client.(*bitbucketv1.APIClient)
+	if !ok {
+		return nil, fmt.Errorf("Invalid Bitbucket Data Center/Server client")
+	}
+
+	prList, err := bitbucketDCClient.DefaultApi.GetPullRequests(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	pullRequest, err := bitbucketv1.GetPullRequestsResponse(prList)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pr := range pullRequest {
+		var repoUrl string
+		for _, link := range pr.FromRef.Repository.Links.Clone {
+			if link.Name == "https" {
+				repoUrl = link.Href
+				break
+			}
+		}
+
+		response = append(response, &GitPullRequest{
+			Name:            pr.Title,
+			Branch:          pr.FromRef.DisplayID,
+			Sha:             pr.FromRef.LatestCommit,
+			SourceRepoId:    pr.FromRef.Repository.Slug,
+			SourceRepoUrl:   repoUrl,
+			SourceRepoOwner: pr.FromRef.Repository.Owner.Name,
+			SourceRepoName:  pr.FromRef.Repository.Name,
+		})
+	}
+
+	return response, nil
+}
+
+func (g *BitbucketServerGitProvider) GetUser() (*GitUser, error) {
+	client := g.getApiClient()
+
+	bitbucketDCClient, ok := client.(*bitbucketv1.APIClient)
+	if !ok {
+		return nil, fmt.Errorf("Invalid Bitbucket Data Center/Server client")
+	}
+
+	// Since BitbucketServer or gfleury/go-bitbucket-v1 doesn't offer an endpoint to query the
+	// currently authenticated user, We instead query the '/rest/api/1.0/application-properties' endpoint
+	// which does not put load on the server and then extract the username from the response header.
+	// Refer this developer community comment: https://community.developer.atlassian.com/t/obtain-authorised-users-username-from-api/24422/2
+	res, err := bitbucketDCClient.DefaultApi.GetApplicationProperties()
+	if err != nil {
+		return nil, err
+	}
+
+	username := res.Header.Get("X-Ausername")
+	if username == "" {
+		return nil, fmt.Errorf("X-Ausername header is missing")
+	}
+
+	user, err := bitbucketDCClient.DefaultApi.GetUser(username)
+	if err != nil {
+		return nil, err
+	}
+
+	var userInfo bitbucketv1.User
+	err = mapstructure.Decode(user.Values, &userInfo)
+
+	if err != nil {
+		return nil, err
+	}
+
+	response := &GitUser{}
+	response.Id = fmt.Sprintf("%d", userInfo.ID)
+	response.Username = username
+	response.Name = userInfo.DisplayName
+
+	return response, nil
+}
+
+func (g *BitbucketServerGitProvider) GetLastCommitSha(staticContext *StaticGitContext) (string, error) {
+	client := g.getApiClient()
+
+	branch := ""
+	if staticContext.Branch != nil {
+		branch = *staticContext.Branch
+	}
```

Comment:
`staticContext.Sha` is missing as a param to the request.
Example in the bitbucket provider:
```golang
	include := ""
	if staticContext.Sha != nil {
		include = *staticContext.Sha
	}

	commits, err := client.Repositories.Commits.GetCommits(&bitbucket.CommitsOptions{
		Owner:       staticContext.Owner,
		RepoSlug:    staticContext.Id,
		Branchortag: branch,
		Include:     include,
	})
```

It's important to use the sha to validate it's integrity. If there is no validation, users could pass invalid commit SHAs into the repo URL.

Reply:
done..